add kubernetes access tests

Ajarmar opened the pull request:
<!-- Choose your PR title carefully as it will be used as the entry in the changelog! -->

<!-- markdownlint-disable MD041 -->
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       <!-- This PR adds a new feature -->
- [ ] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [x] kind/other         <!-- This PR does something else -->

_Optional_: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change   <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change     <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security       <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr](set-me\) <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

<!-- Additional information with kind/admin-change
### Platform Administrator notice
...
-->

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

<!-- Add description of the change -->
Adds end-to-end tests for Kubernetes access.

There are a few things of note here:

- ~~I implemented a cypress test for doing a static e-mail login, and I tried to do the same for "cluster admin" login but I'm not sure it will be possible. Since cypress doesn't have the cookies from your main browser, it requires you to actually log in via your IdP (e.g. Google), which is challenging because:~~
  - ~~We don't know what IdP the user is using,~~
  - ~~Even if we assume Google as IdP, you need to do a real login with your real Google account, which is hard to cover in a test,~~
  - ~~Even if we _could_ cover it in a test, I'm not sure we _should_ go through an OAuth authentication in a test since it's the sort of thing that is meant to be done by a human so you know what access you are granting to the application in question.~~
  - ~~Regardless, if someone has a good idea for how it could be done, I'm open for suggestions, but it seems hard.~~
- The above point has been remedied by using a second static dex user instead of logging in via IdP.
- I tried using the `KUBECACHEDIR` environment variable to change the cache for kubelogin, but I found that kubelogin actually doesn't use that environment variable at all (it will change the cache directory for other kubectl things, but not for kubelogin). It seems that the only way to change the cache directory for kubelogin is via a flag when you run the `oidc-login` command, so I found the easiest implementation of this to be to make a copy of the kubeconfig with that flag added and use it as a "testing" kubeconfig, one for the "admin" login and one for the "static" login". See the `with_test_kubeconfig` functions below.

<!-- Add all issues that are fixed by this PR, use "Part of" instead of "Fixes" if you want to keep issues open. -->
- Fixes elastisys/ck8s-issue-tracker#228

#### Information to reviewers

How to test:

```bash
$ export CK8S_CONFIG_PATH=...
$ make run-end-to-end/kubernetes
```

Some notes about the tests: the cypress tests are slightly flaky, they're implemented in a way that doesn't really conform with cypress best practices (starting a web server inside of the test). Unfortunately this seemed unavoidable because our "web server" in this case is started by running a kubectl command.

They usually work, but I noticed when I had a lot of programs/tabs open and my laptop would be a bit slow running the tests, it was more likely that the tests would fail (usually with an `ECONNREFUSED` error).

Also, you need to make sure that you have decrypted something with your GPG key recently so it doesn't try to prompt you for your passcode when you run the tests (but I think this issue exists for all tests that access the `secrets.yaml` file).

The following config is necessary for the tests:

- `clusterAdmin.users` has `admin@example.com`
- `user.adminUsers` has `dev@example.com`
- For the `authentication-dev` cypress test, you need to have `extraStaticLogins` set, see [this commit](https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/753d1c6b5ac575244c931cdfb380a377ed3930dd). You need to add a `password` field to it in your `secrets.yaml` file; the default password that matches the hash is `password`.

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [x] Proper commit message prefix on all commits
    <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to applications running in all clusters
    - apps sc: changes to applications running in service clusters
    - apps wc: changes to applications running in workload clusters
    - bin: changes to management binaries
    - config: changes to configuration
    - docs: changes to documentation
    - release: release related
    - scripts: changes to scripts
    - tests: changes to tests
    --->
- Change checks:
    - [x] The change is transparent
    - [ ] The change is disruptive
    - [ ] The change requires no migration steps
    - [ ] The change requires migration steps
    - [ ] The change updates CRDs
    - [ ] The change updates the config _and_ the schema
- Documentation checks:
    - [x] The [public documentation](https://github.com/elastisys/welkin) required no updates
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required an update - [link to change](set-me\)
- Metrics checks:
    - [ ] The metrics are still exposed and present in Grafana after the change
    - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts required no updates)
    - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts required an update)
- Logs checks:
    - [ ] The logs do not show any errors after the change
- PodSecurityPolicy checks:
    - [ ] Any changed Pod is covered by Kubernetes Pod Security Standards
    - [ ] Any changed Pod is covered by Gatekeeper Pod Security Policies
    - [ ] The change does not cause any Pods to be blocked by Pod Security Standards or Policies
- NetworkPolicy checks:
    - [ ] Any changed Pod is covered by Network Policies
    - [ ] The change does not cause any dropped packets in the NetworkPolicy Dashboard
- Audit checks:
    - [ ] The change does not cause any unnecessary Kubernetes audit events
    - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
    - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
    - [ ] The bug fix is covered by regression tests

aarnq commented:
Aight, thought: Should we optionally expose a second user through dex (dex@example.com) as an alternative to bind to dev roles? Then we could test access without reliance on external IdPs.

Ajarmar commented:
> Aight, thought: Should we optionally expose a second user through dex (`dev@example.com`) as an alternative to bind to dev roles? Then we could test access without reliance on external IdPs.

Sure, if that covers the QA cases well enough then let's go with that.

Ajarmar commented on tests/end-to-end/kubernetes/bin-access.bats:
```diff
@@ -0,0 +1,55 @@
+@test "static user can list access" {
+  with_test_kubeconfig wc static
+  clear_kubeconfig_cache static
+  echo "Go to http://localhost:8000 and log in with static Email user" >&3
```
I tried using the cache from the cypress tests again and now it works fine. Unfortunately I'm not sure why, but at least it works.

aarnq commented:
> @aarnq I added the ability to create extra static login users (still working on getting the tests updated to use this). I made it so that you can configure the user however you like rather than hard coding it to `dev@example.com`. I'm thinking the tests _will_ be hard coded to `dev@example.com`, so when running the tests, the user would have to add an extra static login user with that specific email. Do you think that way of implementing it is alright?

I like that flexibility :+1: We should probably improve our pre-flight checks to be more clear, and document that expectation in the tests readme, as we already have expectation issues with the trailing dot thing for Grafana, and this might become similar. :sweat_smile:

Ajarmar commented:
> > @aarnq I added the ability to create extra static login users (still working on getting the tests updated to use this). I made it so that you can configure the user however you like rather than hard coding it to `dev@example.com`. I'm thinking the tests _will_ be hard coded to `dev@example.com`, so when running the tests, the user would have to add an extra static login user with that specific email. Do you think that way of implementing it is alright?
> 
> I like that flexibility 👍 We should probably improve our pre-flight checks to be more clear, and document that expectation in the tests readme, as we already have expectation issues with the trailing dot thing for Grafana, and this might become similar. 😅

Great! I've updated the tests to use `dev@example.com`, so the PR is ready for a full review.

rarescosma commented on tests/cypress.config.js:
```diff
@@ -4,6 +4,27 @@ module.exports = defineConfig({
+          return new Promise((resolve, reject) => {
+            process.env.KUBECONFIG = kubeconfig
+            var scriptOutput = "";
+            var child = spawn("kubectl", ["get", "node"],
```
If this is meant to trigger a `kubectl` login, would it be possible to avoid additional API calls, by using something like `"kubectl", ["auth", "whoami"]` instead?

rarescosma reviewed (commented):
Trying to run this test suite in separation using local bats, and the last 3 tests are always failing for me:

```
# bats -T -j 1 end-to-end/kubernetes
authentication-admin.gen.bats
   cypress run: /home/karelian/src/esys/compliantkubernetes-apps/tests/end-to-end/kubernetes/authentication-admin.cy.js
 ✓ kubernetes authentication can login via static dex user [84]
authentication-dev.gen.bats
   cypress run: /home/karelian/src/esys/compliantkubernetes-apps/tests/end-to-end/kubernetes/authentication-dev.cy.js
 ✓ kubernetes authentication can login via extra static dex user [89]
bin-access.bats
 ✓ static user can list access [347]
If cypress auth tests were run previously, test should continue automatically. Otherwise, go to http://localhost:8000 and log in with static Email user dev@example.com
 ✗ static user can delegate admin access [357]
   tags: kubernetes
   (from function `assert_success' in file /usr/local/lib/bats/assert/src/assert_success.bash, line 42,
    in test file end-to-end/kubernetes/bin-access.bats, line 29)
     `assert_success' failed

   -- command failed --
   status : 1
   output : Error from server (Forbidden): rolebindings.rbac.authorization.k8s.io "extra-workload-admins" is forbidden: User "dev@example.com" cannot get resource "rolebindings" in API group "rbac.authorization.k8s.io" in the namespace "staging"
   --

 ✗ static user can delegate view access [355]
   tags: kubernetes
   (from function `assert_success' in file /usr/local/lib/bats/assert/src/assert_success.bash, line 42,
    in test file end-to-end/kubernetes/bin-access.bats, line 37)
     `assert_success' failed

   -- command failed --
   status : 1
   output : Error from server (Forbidden): clusterrolebindings.rbac.authorization.k8s.io "extra-user-view" is forbidden: User "dev@example.com" cannot get resource "clusterrolebindings" in API group "rbac.authorization.k8s.io" at the cluster scope
   --

 ✗ static user cannot run pod as root [238]
   tags: kubernetes
   (in test file end-to-end/kubernetes/bin-access.bats, line 45)
     `kubectl apply -n staging -f "${APPS_PATH}/tests/end-to-end/kubernetes/allow-root-nginx.yaml"' failed
   error: the path "/tests/end-to-end/kubernetes/allow-root-nginx.yaml" does not exist
   error: the path "/tests/end-to-end/kubernetes/allow-root-nginx.yaml" does not exist

6 tests, 3 failures in 19 seconds
```

Also, dex login windows (in my local browser) keep popping up faster than I can login 🤷🏼

aarnq commented:
> Trying to run this test suite in separation using local bats

Note that local runs isn't really supported, it has been optimised to run in the test container.
Some variables might be unset if you do it locally, such as `$APPS_PATH` which is probably why the last test fails due to a missing file.

Ajarmar commented on config/secrets.yaml:
```diff
@@ -171,6 +171,11 @@ dex:
     #    redirectURIs:
     #      - 'http://192.168.42.219:31850/oauth2/callback'
+  extraStaticLogins: []
+    # - email: dev@example.com
+    #   hash: $2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W
+    #   username: dev
+    #   userID: 08a8684b-db88-4b73-90a9-3cd1661f5467
```
Only issue I see with this is that this whole object is applied [in the dex values](https://github.com/elastisys/compliantkubernetes-apps/pull/2548/files#diff-dbe845fd9d8753e5be36f2ae5c74811df69cb738915849c8c902fb890be0a899), so the unhashed password would be stored in the cluster. We could make it into a separate list, something like `extraStaticPasswordsNotHashed`. Thoughts?

aarnq replied:
Or you could use `pick $dict "key1" "key2" ...` in the templates to not get it into there.

Ajarmar replied:
I like that solution. I implemented it with `omit`, see https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/d73d743f89b0ccb1c50e7192d5454f3683de81ca

Ajarmar commented:
> Thought, as the xdg-open works (at least for me) when there hasn't been past authentications, maybe we should consider printing a sops command to easily extract the password as well?

Sorry, I don't fully follow. Where are you suggesting the sops command would be printed?

Ajarmar commented on tests/bats.lib.bash:
```diff
@@ -106,6 +106,33 @@ with_kubeconfig() {
 }
 
+# sets the kubeconfig to use
+# usage: with_kubeconfig <cluster> <static|admin>
```
https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/219b980801883ecaded2318905077c60dd92fd5a

aarnq commented:
> > Thought, as the xdg-open works (at least for me) when there hasn't been past authentications, maybe we should consider printing a sops command to easily extract the password as well?
> 
> Sorry, I don't fully follow. Where are you suggesting the sops command would be printed?

I was thinking with the `Otherwise, go to http://localhost:8000 and log in with static email user dev@example.com` as a way to get the password, but I realise you didn't store the password there so it doesn't really apply.

rarescosma commented:
> Note that local runs isn't really supported, it has been optimised to run in the test container.
> Some variables might be unset if you do it locally, such as `$APPS_PATH` which is probably why the last test fails due to a missing file.

Thanks! Setting `APPS_PATH` and ensuring the `hash: ` key is present in the `sops.yaml` file (I had deleted it for some reason), fixed all the tests, even with local run.

Zash reviewed (commented):
Looks good.

Ajarmar commented on tests/end-to-end/kubernetes/bin-access.bats:
```diff
@@ -0,0 +1,55 @@
+  with_test_kubeconfig wc static-dev
+  run kubectl auth whoami
+  assert_output --partial "dev@example.com"
+  kubectl delete -n staging -f "${APPS_PATH}/tests/end-to-end/kubernetes/allow-root-nginx.yaml" || true
```
Yeah, the only error I was ignoring there was if the resources don't exist, so this is cleaner. Thanks! https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/4850b3a5a97592b38c15540ef4a229a4b8bf52a7

Ajarmar commented on tests/end-to-end/kubernetes/bin-access.bats:
```diff
@@ -0,0 +1,55 @@
+  while [[ ${reason} == "ContainerCreating" ]] || [[ ${reason} == "" ]]; do
+    sleep 1
+    reason=$(kubectl -n staging get pod root-nginx -o=jsonpath='{.status.containerStatuses[0].state.waiting.reason}')
+  done
```
That's a much cleaner solution, thanks. https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/0ef8090ba455f52123e9b3e5f3e9e0b841a5d69f

rarescosma commented on tests/bats.lib.bash:
```diff
@@ -106,6 +106,33 @@ with_kubeconfig() {
+  fi
+
+  BASE_KUBECONFIG="${CK8S_CONFIG_PATH}/.state/kube_config_$1.yaml"
+  export KUBECONFIG="${CK8S_CONFIG_PATH}/.state/kube_config_$1_$2.yaml"
```
Well, it's just a _tad_ bit of an annoyance if you've got aliases to select the current kubeconfig that list everything in the `~/.ck8s/<subfolders>/.state`. 

I'd go for "remove them in a teardown function".

Ajarmar replied:
https://github.com/elastisys/compliantkubernetes-apps/pull/2548/commits/360494ce237f056edb50a072963b2cffcf18d1f3

(I didn't make it delete the kubeconfig in the `authentication-dev` cypress test since we still want that kubeconfig for the bats tests. It's deleted after the bats tests are done.)

anders-elastisys replied:
Ah right I did not consider that, then ignore that option :sweat_smile:

aarnq commented on config/secrets.yaml:
```diff
@@ -175,6 +175,7 @@ dex:
     # - email: dev@example.com
     #   userID: 08a8684b-db88-4b73-90a9-3cd1661f5467
     #   username: dev
+    #   password: password
```
Or the password field can be used to store the password in plain text, and it will not be applied to the environment. With the current comment there is not even a mention of what password fields is.
